recipes/opencv/4.x/conanfile.py

```diff
@@ -284,6 +301,17 @@ def add_components(components):
                 if self.settings.os == "Linux":
                     self.cpp_info.components[conan_component].system_libs = ["dl", "m", "pthread", "rt"]
 
+                if self.settings.os == "Android":
+                    self.cpp_info.components[conan_component].includedirs = [
+                        os.path.join("sdk", "native", "jni", "include")]
+                    self.cpp_info.components[conan_component].system_libs += ["log"]
+                    if not self.options.shared:
+                        self.cpp_info.components[conan_component].libdirs.append(
+                            os.path.join("sdk", "native", "staticlibs", tools.to_android_abi(str(self.settings.arch))))
```

Comment:
For some reason the OpenCV build on my system omits the architecture folder and uses `staticlibs` directly. Can we have both?

Reply:
I'm on Ubuntu 20.04, and we use `android_ndk_installer/r20@bincrafters/stable`, one revision older than yours. I have a pure cmake/make project, which uses my own helper functions to install/update conan packages defined in a project specific `conanfile.py`. And we use a profile file for android:

```ini
include(default)

[settings]
# TODO multiple archs? armv7, armv7s, arm64, x86_64, etc?
arch=armv8
compiler=clang
compiler.libcxx=libc++
compiler.version=8
os=Android
os.api_level=21

[build_requires]
android_ndk_installer/r20@bincrafters/stable

[options]
[env]
```

Thats all, no cmake toolchain file as far as I know. I have no idea what's the issue with `CMAKE_SYSTEM_PROCESSOR`? 

> For some reason the OpenCV build on my system omits the architecture folder and uses staticlibs directly. Can we have both?

Hm, I think before adding extra include directories we should try to understand why this happens? Maybe some flaws in the configuration we should fix instead?

In general, I think we should do this in a separate step, otherwise we get confused. What do you think?

Reply:
I already saw this behavior in the bincrafters recipe. IIRC the problem is related to Multiarchitecture ndk vs single architecture but it has been a while.  I believe it was triggered by some cmake variable beeing empty in opencv.

Reply:
would be nice if you can remember which variables influences this, than we can define the path correctly 😏

Reply:
Sry, was on my phone. I looked into OpenCV's cmake now and I'd say the problem is in [`cmake/OpenCVInstallLayout.cmake`](https://github.com/opencv/opencv/blob/4.5.0/cmake/OpenCVInstallLayout.cmake#L10) which defines:

```
ocv_update(OPENCV_LIB_ARCHIVE_INSTALL_PATH    "sdk/native/staticlibs/${ANDROID_NDK_ABI_NAME}")
```

Which in turn is used in `cmake/OpenCVModule.cmake` for `macro(_ocv_create_module)`, setting the `ARCHIVE DESTINATION`:

```
ocv_install_target(${the_module} EXPORT OpenCVModules OPTIONAL
      RUNTIME DESTINATION ${OPENCV_BIN_INSTALL_PATH} COMPONENT libs
      LIBRARY DESTINATION ${OPENCV_LIB_INSTALL_PATH} COMPONENT libs NAMELINK_SKIP
      ARCHIVE DESTINATION ${OPENCV_LIB_ARCHIVE_INSTALL_PATH} COMPONENT dev
      )
```

`ANDROID_NDK_ABI_NAME` is defined in `platforms/android/android.toolchain.cmake` (just a copy of the default android toolchain file), which I presume not to be loaded in my case. I don't see it being included by OpenCV's CMake scripts.

Reply:
I did some "prinf debugging" (well, ok, more "message" debugging inside cmake files). I can see that at the beginning of OpenCV's main CMakeLists.txt `ANDROID_NDK_ABI_NAME` is set to `arm64-v8a` with my android profile in use. Furthermore, I found out that it doesn't come from OpenCV's `android.toolchain.cmake`, which seems not to be used at all. But instead it comes from the same file in the `android_ndk_installer`! (r20 in my case).

SO, AFAIK it looks like this variable should always be set, otherwise it looks like some error to me. The update of the `OPENCV_LIB_ARCHIVE_INSTALL_PATH` as you show above seems to assume an always set value for `ANDROID_NDK_ABI_NAME` I would say.

Reply:
I looked into that a bit more and noticed that the problem only occured when using the cmake wrapper instead of the profile. The problem was that I did not specify the NDK as build_requires. I assume conan then just picked up the system NDK and did somehow ignore their toolchain file.

> I can see that at the beginning of OpenCV's main CMakeLists.txt ANDROID_NDK_ABI_NAME is set to arm64-v8a

There is nothing in OpenCV CMakeLists.txt that does this. The most likely behavior is that CMake itself loads the android toolchain file (probably instructed to do so by conan).

Reply:
> There is nothing in OpenCV CMakeLists.txt that does this. The most likely behavior is that CMake itself loads the android toolchain file (probably instructed to do so by conan).

Well, using the the android profile specifying an android_ndk_installer is one way to get this variable from this wrapper. This is the way my project works. If you are using CMakeLists.txt directly, I guess you have to specify the toolchain file manually at the command line. I remeber we had to use this years ago for some embedded /cross compile projects too.

I think I mark this as resolved, don't know if there is some action required for this, ok?

Reply:
Okay from my end.